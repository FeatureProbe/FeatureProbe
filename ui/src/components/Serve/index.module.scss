.serve {
  width: 600px;
}

.serve-select {
  display: flex;
  margin-bottom: 12px;
  align-items: center;
}

.serve-text {
  width: 52px;
  font-size: 13px;
  color: #212529;
}

.label-required {
  color: #ff5f57;
  font-size: 20px;
  vertical-align: middle;
  margin-right: 4px;
  height: 16px;
  display: inline-block;
}

.serve-form-field {
  width: 328px;
}

.serve-dropdown {
  width: 316px;
}

.percentage-content {
  margin-left: 52px;
}

.percentage {
  display: flex;
  height: 14px;
  padding: 3px;
  border-radius: 3px;
  border: 1px solid rgba(206, 212, 218, 0.7);
  margin-bottom: 8px;
}

.percentage-item {
  display: inline-block;
}

.message {
  width: 100%;
  height: 27px;
  background: #fff5f4;
  border-radius: 6px;
  line-height: 27px;
  font-size: 12px;
  color: #592d00;
  padding: 0 14px;
}

.message-iconfont {
  color: #f5483b;
  vertical-align: middle;
  margin-right: 4px;
}

.variation {
  margin-top: 8px;
}

.variation-item {
  display: flex;
  align-items: center;
  margin: 12px 0;

  :global .ui.icon.input > input {
    padding-right: 50px !important;
  }
}

.variation-item-input {
  width: 100px !important;
}

.input-percentage {
  position: absolute;
  right: 2px;
  top: 2px;
  font-size: 13px;
  color: #495057;
  width: 24px;
  height: 23px;
  background: #ffffff;
  border-radius: 6px;
  text-align: center;
  line-height: 24px;
}

.input-operation {
  display: inline-flex;
  flex-direction: column;
  position: absolute;
  right: 30px;
  padding-top: 1px;
}

.iconfont {
  font-size: 12px !important;
  height: 10px;
  cursor: pointer;
  position: relative;
  top: -2px;
}

.variation-name {
  margin-left: 10px;
  display: inline-flex;
  align-items: center;
}

.variation-name-text {
  display: inline-block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  max-width: 420px;
}

.variation-name-color {
  display: inline-block;
  width: 12px;
  height: 12px;
  margin-right: 5px;
  border-radius: 4px;
}

.dropdown-variation {
  display: flex;
  align-items: center;
}

.dropdown-variation-color {
  display: inline-block;
  width: 12px;
  height: 12px;
  margin-right: 5px;
  border-radius: 4px;
}

.dropdown-variation-text {
  display: inline-block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  flex: 1;
}

.angle-down {
  position: absolute;
  right: 10px;
  color: #74788d;
  top: 8px;
}

.error-text {
  color: #f46a6a;
  font-size: 12px;
}