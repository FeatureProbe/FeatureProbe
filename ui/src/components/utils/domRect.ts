export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}