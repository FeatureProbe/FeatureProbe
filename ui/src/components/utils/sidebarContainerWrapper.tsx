import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
