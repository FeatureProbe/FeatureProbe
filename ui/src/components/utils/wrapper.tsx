import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
