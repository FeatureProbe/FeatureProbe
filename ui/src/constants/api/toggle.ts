import { origin } from './constant';

const ToggleURI = {
  getToggleListURI: `${origin}/projects/:projectKey/toggles`,
  getToggleInfoURI: `${origin}/projects/:projectKey/toggles/:toggleKey`,
  environmentURI: `${origin}/projects/:projectKey/environments`,
  targetingURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/targeting`,
  approvalTargetingURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/targeting/approval`,
  targetingDiffURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/targeting/diff`,
  createToggleURI: `${origin}/projects/:projectKey/toggles`,
  editToggleURI: `${origin}/projects/:projectKey/toggles/:toggleKey`,
  offlineToggleURI: `${origin}/projects/:projectKey/toggles/:toggleKey/offline`,
  restoreToggleURI: `${origin}/projects/:projectKey/toggles/:toggleKey/restore`,
  tagsURI: `${origin}/projects/:projectKey/tags`,
  trafficURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/traffic`,
  toggleAccessURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/access`,
  toggleTrackEventURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/metric/status`,
  toggleExistURI: `${origin}/projects/:projectKey/toggles/exists`,
  environmentExistURI: `${origin}/projects/:projectKey/environments/exists`,
  targetingVersionsURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/targeting/versions`,
  targetingVersionsByVersionURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/targeting/versions/:version`,
  getAttributesURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/targeting/attributes`,
  getPrerequisiteToggleURI: `${origin}/projects/:projectKey/environments/:environmentKey/toggles/:toggleKey/targeting/prerequisite/toggles`,
};

export default ToggleURI;
