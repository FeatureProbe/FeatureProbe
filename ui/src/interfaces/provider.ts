/* eslint-disable @typescript-eslint/no-explicit-any */
import { ContainerProviderProps } from 'unstated-next';
import { ReactElement } from 'react';

interface IFn {
  (): any;
}
  
export interface IContainer {
  Provider: React.ComponentType<ContainerProviderProps<void>>;
  useContainer: IFn;
}

export interface IChildren {
  children: ReactElement
}