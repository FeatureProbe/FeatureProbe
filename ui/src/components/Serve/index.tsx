import { useEffect, useState, SyntheticEvent, ChangeEvent, useCallback } from 'react';
import { Dropdown, Input, DropdownItemProps, DropdownProps, Form } from 'semantic-ui-react';
import { FormattedMessage, useIntl } from 'react-intl';
import { cloneDeep } from 'lodash';
import Icon from 'components/Icon';
import { IServe, IVariation } from 'interfaces/targeting';
import { IContainer } from 'interfaces/provider';
import { VariationColors } from 'constants/colors';
import { useFormErrorScrollIntoView } from 'hooks';
import styles from './index.module.scss';

interface IAttr {
  [key: string]: string;
}
interface IProps {
  index?: number;
  id?: string;
  disabled?: boolean;
  variations: IVariation[];
  serve?: IServe;
  handleChangeServe: (item: IServe) => void;
  hooksFormContainer: IContainer;
  customStyle?: IAttr;
}

type IDropItemProps = Partial<DropdownItemProps>;

const TOTAL = 100;
const MIN = 0;

const Serve = (props: IProps) => {
  const {
    id,
    index,
    serve,
    disabled,
    variations,
    customStyle,
    hooksFormContainer,
    handleChangeServe,
  } = props;

  const [ value, saveValue ] = useState<string | number | boolean | (string | number | boolean)[] | undefined>(-1);
  const [ percentageShow, setPercentageShow ] = useState<boolean>(false);
  const [ variationsInUse, setVariationsInUse ] = useState<IVariation[]>([]);
  const [ variationsOptions, setVariatiosOptions ] = useState<IDropItemProps[]>([]);
  const [ total, setTotal ] = useState<number>(0);
  const intl = useIntl();

  const {
    formState: { errors },
    setValue,
    trigger,
    register,
    setError,
    clearErrors,
  } = hooksFormContainer.useContainer();
  const { registerErrorName } = useFormErrorScrollIntoView();

  useEffect(() => {
    setVariationsInUse(cloneDeep(variations));
  }, [variations]);

  useEffect(() => {
    const options: IDropItemProps[] = variationsInUse.map((item: IDropItemProps, index: number) => {
      return {
        value: index,
        text: item.name || item.value,
        content: (
          <div className={styles['dropdown-variation']}>
            <span style={{background: VariationColors[index % 20]}} className={styles['dropdown-variation-color']}></span>
            <span className={styles['dropdown-variation-text']}>
              { item.name || item.value }
            </span>
          </div>
        ),
      };
    });

    options.push({
      text: intl.formatMessage({id: 'targeting.serve.percentage.rollout'}),
      value: variationsInUse.length,
    });

    setVariatiosOptions(options);
  }, [intl, variationsInUse]);

  useEffect(() => {
    if (serve?.split) {
      setPercentageShow(true);
      saveValue(variationsInUse.length);
      let total = 0;
      variationsInUse?.forEach((variation: IVariation, index: number) => {
        variation.inputValue = (serve.split?.[index] ?? 0) / TOTAL;
        variation.percentage = variation.inputValue + '%';
        total += Number(variation.inputValue);
      });
      setTotal(total);
    } else if (serve?.select !== undefined) {
      setPercentageShow(false);
      if (serve.select >= variationsInUse.length) {
        saveValue(-1);
        setValue(id ? `rule_${id}_serve` : 'defaultServe', null);
        setError(id ? `rule_${id}_serve` : 'defaultServe');
      } else {
        saveValue(serve.select);
        setValue(id ? `rule_${id}_serve`: 'defaultServe', serve.select);
        clearErrors(id ? `rule_${id}_serve`: 'defaultServe');
      }
    } else {
      setPercentageShow(false);
      saveValue(-1);
    }
  }, [id, index, serve, variationsInUse, setValue, setError, clearErrors]);

  useEffect(() => {
    if (total !== TOTAL && percentageShow) {
      setError(id ? `rule_${id}_serve_total` : 'defaultServe_total', { message: 'total is not 100%' });
    } else {
      clearErrors(id ? `rule_${id}_serve_total`: 'defaultServe_total');
    }
  }, [id, total, index, percentageShow, setError, clearErrors]);

  const handleChange = useCallback((event: SyntheticEvent, data: DropdownProps) => {
    if (data.value === variations.length) {
      setPercentageShow(true);
      const split = variationsInUse.map((variation) => {
        return Number(variation.inputValue || 0) * TOTAL;
      });
      handleChangeServe({
        split,
      });
    } else {
      setPercentageShow(false);
      handleChangeServe({
        select: Number(data.value),
      });
    }
    saveValue(data.value);
    setTotal(0);
  }, [variations.length, variationsInUse, handleChangeServe]);

  const changePercentage = useCallback((index: number, value: number) => {
    if (value < 0) {
      value = 0;
    } else if (value > TOTAL) {
      value = TOTAL;
    }

    let total = 0;
    variationsInUse[index].inputValue = value;

    if (variationsInUse.length == 2) {
      // auto adjustment to keep the total rollout is 100%
      variationsInUse[1 - index].inputValue = TOTAL - value;
    }

    variationsInUse.forEach(variation => {
      if (variation.inputValue) {
        total += Number(variation.inputValue);
      }
    });

    variationsInUse.forEach(variation => {
      if (variation.inputValue) {
        if (total > TOTAL) {
          variation.percentage = Number(variation.inputValue) / total * TOTAL + '%';
        } else {
          variation.percentage = Number(variation.inputValue) + '%';
        }
      } else {
        variation.percentage = '0%';
      }
    });

    const split = variationsInUse.map((variation) => {
      return Number(variation.inputValue || 0) * TOTAL;
    });

    handleChangeServe({
      split,
    });

    setTotal(total);
  }, [variationsInUse, handleChangeServe]);

  const handleInputChange = useCallback((e: ChangeEvent, index: number) => {
    changePercentage(index, Number((e.target as HTMLInputElement).value) || MIN);
  }, [changePercentage]);

  const handleMinus = useCallback((index: number) => {
    const value = Number(variationsInUse[index].inputValue);
    if (value <= MIN || disabled) {
      return;
    }
    changePercentage(index, value - 1);
  }, [disabled, variationsInUse, changePercentage]);

  const handleAdd = useCallback((index: number) => {
    const value = Number(variationsInUse[index].inputValue);
    if (value >= TOTAL || disabled) {
      return;
    }
    changePercentage(index, value + 1);
  }, [disabled, variationsInUse, changePercentage]);

	return (
    <div className={styles.serve}>
      <div className={styles['serve-select']}>
        <span className={styles['serve-text']}>
          <span className={styles['label-required']}>*</span>
          <FormattedMessage id='common.serve.text' />
        </span>
        <Form.Field className={styles['serve-form-field']} style={customStyle}>
          <Dropdown
            placeholder={intl.formatMessage({id: 'targeting.serve.required'})}
            className={styles['serve-dropdown']}
            style={customStyle}
            selection
            floating
            value={value}
            openOnFocus={false}
            disabled={disabled}
            options={variationsOptions}
            error={ errors[id ? `rule_${id}_serve`: 'defaultServe'] ? true : false }
            {
              ...register(id ? `rule_${id}_serve`: 'defaultServe', { 
                required: true,
              })
            }
            onChange={async (e: SyntheticEvent, detail: DropdownProps) => {
              handleChange(e, detail);
              setValue(detail.name, detail.value);
              await trigger(id ? `rule_${id}_serve`: 'defaultServe');
            }}
            icon={<Icon customclass={styles['angle-down']} type='angle-down' />}
          />
          { errors[id ? `rule_${id}_serve`: 'defaultServe'] && <div className={styles['error-text']}>
            { intl.formatMessage({id: 'targeting.serve.required'}) }
          </div> }
        </Form.Field>
      </div>
      {
        percentageShow && (
          <div className={styles['percentage-content']}>
            <div className={styles.percentage}>
              {
                variationsInUse.map((item: IVariation, index: number) => {
                  return (
                    <span 
                      key={item.value + '_' + index}
                      style={{width: item.percentage, background: VariationColors[index % 20]}} 
                      className={styles['percentage-item']}
                    />
                  );
                })
              }
            </div>
            {
              total !== TOTAL && (
                <div {...registerErrorName(id ? `rule_${id}_serve_total` : 'defaultServe_total')} className={styles.message}>
                  <Icon customclass={styles['message-iconfont']} type='remove-circle' />
                  {
                    intl.formatMessage({
                      id: 'targeting.serve.percentage.error.text'
                    }, {
                      value: TOTAL - total
                    })
                  }
                </div>
              )
            }
            <div className={styles.variation}>
              {
                variationsInUse?.map((item: IVariation, index: number) => {
                  return (
                    <div key={item.id} className={styles['variation-item']}>
                      <Input 
                        size='mini'
                        className={styles['variation-item-input']}
                        value={item.inputValue ?? ''}
                        disabled={disabled}
                        onChange={(e: ChangeEvent) => handleInputChange(e, index)}
                        icon={
                          <div>
                            <span className={styles['input-percentage']}>%</span>
                            <span className={styles['input-operation']}>
                              <Icon customclass={styles['iconfont']} type='angle-up' onClick={() => handleAdd(index)} />
                              <Icon customclass={styles['iconfont']} type='angle-down' onClick={() => handleMinus(index)} />
                            </span>
                          </div>
                        }
                      />

                      <div className={styles['variation-name']}>
                        <span style={{ background: VariationColors[index % 24] }} className={styles['variation-name-color']}></span>
                        <span className={styles['variation-name-text']}>
                          { item.name || item.value }
                        </span>
                      </div>
                    </div>
                  );
                })
              }
            </div>
          </div>
        )
      }
    </div>
	);
};

export default Serve;