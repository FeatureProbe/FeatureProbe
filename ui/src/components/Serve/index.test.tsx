import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Serve from '.';
import { hooksFormContainer } from 'pages/toggle/provider';
import { IntlWrapper } from 'components/utils/wrapper';
import { IVariation } from 'interfaces/targeting';

const Wrapper: React.FC = (props) => {
  return (
    <IntlWrapper>
      <hooksFormContainer.Provider>{props.children}</hooksFormContainer.Provider>
    </IntlWrapper>
  );
};

const variations: IVariation[] = [
  {
    id: '1',
    value: 'true',
    name: 'Variation 1',
    description: 'This is a description',
  },
  {
    id: '2',
    value: 'false',
    name: 'Variation 2',
    description: 'This is a description',
  },
];

it('Serve snapshot', (done) => {
  (async () => {
    const { asFragment, rerender } = render(
      <Serve variations={[]} handleChangeServe={jest.fn()} hooksFormContainer={hooksFormContainer} />,
      {
        wrapper: Wrapper,
      }
    );
    expect(asFragment()).toMatchSnapshot();

    rerender(<Serve variations={variations} handleChangeServe={jest.fn()} hooksFormContainer={hooksFormContainer} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(
      <Serve
        disabled={true}
        variations={variations}
        serve={{ split: [0, 0], select: 2 }}
        handleChangeServe={jest.fn()}
        hooksFormContainer={hooksFormContainer}
      />
    );
    expect(asFragment()).toMatchSnapshot();

    rerender(
      <Serve
        disabled={false}
        variations={variations}
        serve={{ select: 0 }}
        handleChangeServe={jest.fn()}
        hooksFormContainer={hooksFormContainer}
      />
    );
    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});

test('Serve select', (done) => {
  (async () => {
    const mockHandleChangeServe = jest.fn();

    const { asFragment } = render(
      <Serve
        variations={variations}
        handleChangeServe={mockHandleChangeServe}
        hooksFormContainer={hooksFormContainer}
      />,
      {
        wrapper: Wrapper,
      }
    );
    const selectEle = screen.getByText('Please select serve');
    await userEvent.click(selectEle);
    const options = screen.getAllByRole('option');
    await userEvent.click(options[0]);

    expect(mockHandleChangeServe).toHaveBeenLastCalledWith({ select: 0 });

    await userEvent.click(selectEle);
    await userEvent.click(options[2]);

    expect(mockHandleChangeServe).toHaveBeenLastCalledWith(expect.objectContaining({ split: [0, 0] }));
    expect(asFragment()).toMatchSnapshot();

    const textInputs = screen.getAllByRole('textbox');
    await userEvent.click(textInputs[0]);
    await userEvent.keyboard('50');
    await userEvent.click(textInputs[1]);
    await userEvent.keyboard('60');

    await userEvent.click(textInputs[0]);
    await userEvent.keyboard('{ctrl}{a}');

    await userEvent.keyboard('1');
    const down = document.querySelector('span.input-operation>.icon-angle-down');
    const up = document.querySelector('span.input-operation>.icon-angle-up');

    down && await userEvent.click(down);
    down && await userEvent.click(down);
    await userEvent.click(textInputs[0]);
    await userEvent.keyboard('{ctrl}{a}');
    await userEvent.keyboard('99');
    up && await userEvent.click(up);
    up && await userEvent.click(up);
    await userEvent.click(textInputs[0]);
    await userEvent.keyboard('{ctrl}{a}');
    await userEvent.keyboard('50');


    await userEvent.click(textInputs[1]);
    await userEvent.keyboard('{ctrl}{a}');

    await userEvent.keyboard('50');
    expect(mockHandleChangeServe).toHaveBeenLastCalledWith(expect.objectContaining({ split: [0, 10000] }));

    done();
  })();
});
