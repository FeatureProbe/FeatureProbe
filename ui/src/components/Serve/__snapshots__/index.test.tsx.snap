// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serve select 1`] = `
<DocumentFragment>
  <div
    class="serve"
  >
    <div
      class="serve-select"
    >
      <span
        class="serve-text"
      >
        <span
          class="label-required"
        >
          *
        </span>
        Serve
      </span>
      <div
        class="field serve-form-field"
      >
        <div
          aria-expanded="false"
          class="ui floating selection dropdown serve-dropdown"
          name="defaultServe"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            a percentage rollout
          </div>
          <span
            class="iconfont icon-angle-down angle-down"
            customclass="angle-down"
            type="angle-down"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(85, 110, 230);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 1
                </span>
              </div>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(255, 156, 27);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 2
                </span>
              </div>
            </div>
            <div
              aria-checked="true"
              aria-selected="true"
              class="active selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                a percentage rollout
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="percentage-content"
    >
      <div
        class="percentage"
      >
        <span
          class="percentage-item"
          style="background: rgb(85, 110, 230);"
        />
        <span
          class="percentage-item"
          style="background: rgb(255, 156, 27);"
        />
      </div>
      <div
        class="message"
        name="defaultServe_total"
      >
        <span
          class="iconfont icon-remove-circle message-iconfont"
          customclass="message-iconfont"
          type="remove-circle"
        />
        100% needs to be assigned to a variation
      </div>
      <div
        class="variation"
      >
        <div
          class="variation-item"
        >
          <div
            class="ui mini icon input variation-item-input"
          >
            <input
              type="text"
              value=""
            />
            <div>
              <span
                class="input-percentage"
              >
                %
              </span>
              <span
                class="input-operation"
              >
                <span
                  class="iconfont icon-angle-up iconfont"
                  customclass="iconfont"
                  type="angle-up"
                />
                <span
                  class="iconfont icon-angle-down iconfont"
                  customclass="iconfont"
                  type="angle-down"
                />
              </span>
            </div>
          </div>
          <div
            class="variation-name"
          >
            <span
              class="variation-name-color"
              style="background: rgb(85, 110, 230);"
            />
            <span
              class="variation-name-text"
            >
              Variation 1
            </span>
          </div>
        </div>
        <div
          class="variation-item"
        >
          <div
            class="ui mini icon input variation-item-input"
          >
            <input
              type="text"
              value=""
            />
            <div>
              <span
                class="input-percentage"
              >
                %
              </span>
              <span
                class="input-operation"
              >
                <span
                  class="iconfont icon-angle-up iconfont"
                  customclass="iconfont"
                  type="angle-up"
                />
                <span
                  class="iconfont icon-angle-down iconfont"
                  customclass="iconfont"
                  type="angle-down"
                />
              </span>
            </div>
          </div>
          <div
            class="variation-name"
          >
            <span
              class="variation-name-color"
              style="background: rgb(255, 156, 27);"
            />
            <span
              class="variation-name-text"
            >
              Variation 2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Serve snapshot 1`] = `
<DocumentFragment>
  <div
    class="serve"
  >
    <div
      class="serve-select"
    >
      <span
        class="serve-text"
      >
        <span
          class="label-required"
        >
          *
        </span>
        Serve
      </span>
      <div
        class="field serve-form-field"
      >
        <div
          aria-expanded="false"
          class="ui floating selection dropdown serve-dropdown"
          name="defaultServe"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Please select serve
          </div>
          <span
            class="iconfont icon-angle-down angle-down"
            customclass="angle-down"
            type="angle-down"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="true"
              class="selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                a percentage rollout
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Serve snapshot 2`] = `
<DocumentFragment>
  <div
    class="serve"
  >
    <div
      class="serve-select"
    >
      <span
        class="serve-text"
      >
        <span
          class="label-required"
        >
          *
        </span>
        Serve
      </span>
      <div
        class="field serve-form-field"
      >
        <div
          aria-expanded="false"
          class="ui floating selection dropdown serve-dropdown"
          name="defaultServe"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Please select serve
          </div>
          <span
            class="iconfont icon-angle-down angle-down"
            customclass="angle-down"
            type="angle-down"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="true"
              class="selected item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(85, 110, 230);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 1
                </span>
              </div>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(255, 156, 27);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 2
                </span>
              </div>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                a percentage rollout
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Serve snapshot 3`] = `
<DocumentFragment>
  <div
    class="serve"
  >
    <div
      class="serve-select"
    >
      <span
        class="serve-text"
      >
        <span
          class="label-required"
        >
          *
        </span>
        Serve
      </span>
      <div
        class="field serve-form-field"
      >
        <div
          aria-disabled="true"
          aria-expanded="false"
          class="ui disabled floating selection dropdown serve-dropdown"
          name="defaultServe"
          role="listbox"
          tabindex="-1"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            a percentage rollout
          </div>
          <span
            class="iconfont icon-angle-down angle-down"
            customclass="angle-down"
            type="angle-down"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(85, 110, 230);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 1
                </span>
              </div>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(255, 156, 27);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 2
                </span>
              </div>
            </div>
            <div
              aria-checked="true"
              aria-selected="true"
              class="active selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                a percentage rollout
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="percentage-content"
    >
      <div
        class="percentage"
      >
        <span
          class="percentage-item"
          style="width: 0%; background: rgb(85, 110, 230);"
        />
        <span
          class="percentage-item"
          style="width: 0%; background: rgb(255, 156, 27);"
        />
      </div>
      <div
        class="message"
        name="defaultServe_total"
      >
        <span
          class="iconfont icon-remove-circle message-iconfont"
          customclass="message-iconfont"
          type="remove-circle"
        />
        100% needs to be assigned to a variation
      </div>
      <div
        class="variation"
      >
        <div
          class="variation-item"
        >
          <div
            class="ui mini disabled icon input variation-item-input"
          >
            <input
              disabled=""
              tabindex="-1"
              type="text"
              value="0"
            />
            <div>
              <span
                class="input-percentage"
              >
                %
              </span>
              <span
                class="input-operation"
              >
                <span
                  class="iconfont icon-angle-up iconfont"
                  customclass="iconfont"
                  type="angle-up"
                />
                <span
                  class="iconfont icon-angle-down iconfont"
                  customclass="iconfont"
                  type="angle-down"
                />
              </span>
            </div>
          </div>
          <div
            class="variation-name"
          >
            <span
              class="variation-name-color"
              style="background: rgb(85, 110, 230);"
            />
            <span
              class="variation-name-text"
            >
              Variation 1
            </span>
          </div>
        </div>
        <div
          class="variation-item"
        >
          <div
            class="ui mini disabled icon input variation-item-input"
          >
            <input
              disabled=""
              tabindex="-1"
              type="text"
              value="0"
            />
            <div>
              <span
                class="input-percentage"
              >
                %
              </span>
              <span
                class="input-operation"
              >
                <span
                  class="iconfont icon-angle-up iconfont"
                  customclass="iconfont"
                  type="angle-up"
                />
                <span
                  class="iconfont icon-angle-down iconfont"
                  customclass="iconfont"
                  type="angle-down"
                />
              </span>
            </div>
          </div>
          <div
            class="variation-name"
          >
            <span
              class="variation-name-color"
              style="background: rgb(255, 156, 27);"
            />
            <span
              class="variation-name-text"
            >
              Variation 2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Serve snapshot 4`] = `
<DocumentFragment>
  <div
    class="serve"
  >
    <div
      class="serve-select"
    >
      <span
        class="serve-text"
      >
        <span
          class="label-required"
        >
          *
        </span>
        Serve
      </span>
      <div
        class="field serve-form-field"
      >
        <div
          aria-disabled="false"
          aria-expanded="false"
          class="ui floating selection dropdown serve-dropdown"
          name="defaultServe"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Variation 1
          </div>
          <span
            class="iconfont icon-angle-down angle-down"
            customclass="angle-down"
            type="angle-down"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="true"
              aria-selected="true"
              class="active selected item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(85, 110, 230);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 1
                </span>
              </div>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <div
                class="text dropdown-variation"
              >
                <span
                  class="dropdown-variation-color"
                  style="background: rgb(255, 156, 27);"
                />
                <span
                  class="dropdown-variation-text"
                >
                  Variation 2
                </span>
              </div>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                a percentage rollout
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
